### ISSUE TITLE
Download new maps asynchronously 

### ISSUE THREAD

**Issue by mokun** [mars-sim/mars-sim#1037]
**Is your feature request related to a problem? Please describe.**
- If running mars-sim for the first time in a new machine, choose a map with resolution higher than level 0 will likely require downloading maps from github. 
- Since the size of the highest resolution of maps can be over 130 MB, Mars Navigator shows a black box and is being uninformative on what's going on. 

**Describe the solution you desire to see**
- Allow downloading a map from the internet asynchronously.
- Have a top level modal window pop up, pause the simulation, and provide a progress bar to show percentage of download.
- Allow the newly downloaded map to be loaded afterward.

**Describe the rationale**
- This gives player a better user experience.

**Specifications  (please complete the following information):**
 - pre 3.7.0 build 8634
 
Note: see #901 regarding maps

**Comment by bevans2000** [comment]
This is not that easy because the "users" of the map data are coded to be synchronous so if there is no map data that can not function. Nevertheless, you are totally correct we need to be doing async downloads, that bit is easy.

Essentially I will change the FileLocator class to return a ```Future<File>``` which will allow us to be asynch underneath and then work on the clients.

**Comment by mokun** [comment]
> This is not that easy because the "users" of the map data are coded to be synchronous so if there is no map data that can not function.

But since we already have those resolution levels defined for each map, it should be easy to switch maps.

**Comment by mokun** [comment]
I'd love to see that, given one single large jpg map, the player may select the resolution level of the map to be downloaded in parallel chunks. 

But this would probably involve some background processing on GitHub prior to the raw download and it won't be easily implemented.

**Comment by bevans2000** [comment]
Also we need to host the large map files in the pages repo as well. The main repo is quite large and we can move these big files.

**Comment by mokun** [comment]
Is it just changing the link to point to the page repo ?

**Comment by bevans2000** [comment]
Yes and moving the actual files.

**Comment by mokun** [comment]
I just read this [article](https://www.kdgregory.com/index.php?page=swing.async) on Asynchronous Operation. 

Hope we work out a solution soon for downloading high resolution maps.

**Comment by mokun** [comment]
Is that true our goal is to put the codes such as the method below into a background thread ?

In FileLocator,
```java
    /**
     * Attempts to find a file in the remote content repository.
     * 
     * @param name Name of the file to locate
     * @return
     * @throws URISyntaxException 
     */
    private static InputStream openRemoteContent(String name) throws URISyntaxException {
        // Check remote content
        String fileURL = contentURL + name;
        try {
            return (new URI(fileURL)).toURL().openStream();
        } catch (IOException e) {
            logger.warning("Problem opening stream: " + e);
            return null;
        }    
    }


```

**Comment by bevans2000** [comment]
Yes but the method needs to use a Future so it callback when the stream is ready. It's down for release 4.0 but we could move it to  3.9

**Comment by mokun** [comment]
I just changed it to 3.9.0.

The benefit of asynchronous download is that mars-sim may load the following data on the fly : 
1. Higher resolution maps
2. Higher elevation data set
3. Sound tracks
4. Future map tiles / regional elevation data

**Comment by bevans2000** [comment]
I agree this is a good one for 3.9. It will reduce the size of the jars.

Let's dropped everything from 3.9 to reset it and then add in our key issues.

**Comment by bevans2000** [comment]
So downloading the file async within ```FileLocator``` is not the problem; that is quite easy. The problem lies in the callers of the FileLocator method; they are blocking awaiting the file contents. So if you look at IntegerMapData.loadMapData, that whole class is dependent upon reading the file to identify resolution for the map. These calling classes need to be async in their logic. So a MapData object might get created via the UI but it is not immediately available. It becomes available in the background.
Hence the starting point is in the map widget has to delay flipping the map layer over until the later is ready. I think I can make it work. Key is to add a ```isReady``` method to the ```MapData``` interface. And also some form of callback method once the File is ready so it can automatically read by the MapData.

**Comment by mokun** [comment]
> Hence the starting point is in the map widget has to delay flipping the map layer over until the later is ready. 

Sure. and there are a few trappings that are worth mentioning.

Currently, updating the map is actually done on an executive thread called *MapTask* in `MapPanel`.

But interestingly one can also run the following line directly in order to draw the map: 

> marsMap.drawMap(centerCoords, rho); 

```java
	/**
	 * Updates the display on a thread.
	 * 
	 * @param rho
	 */
	public void updateDisplay(double rho) {
		if ((desktop.isToolWindowOpen(NavigatorWindow.NAME) 
			|| desktop.isToolWindowOpen(MissionWindow.NAME))
			&& (!executor.isTerminated() || !executor.isShutdown())) {
				executor.execute(new MapTask(rho));
		}
	}

	/**
	 * The task class for drawing the surface map.
	 */
	class MapTask implements Runnable {

		private double rho;
		
		private MapTask(double rho) {
			this.rho = rho;
		}

		@Override
		public void run() {
			try {
				mapError = false;

				if (centerCoords == null) {
					logger.severe("centerCoords is null.");
					centerCoords = new Coordinates(HALF_PI, 0);
				}

				marsMap.drawMap(centerCoords, rho);
				
				// Update the zoom slider value
				updateScaleZoomSlider(rho);
//				if (navwin != null)
//					navwin.updateScaleZoomSlider(rho);
//				else
//					updateScale(rho);
				
				wait = false;
				repaint();
				
			} catch (Exception e) {
				mapError = true;
				mapErrorMessage = e.getMessage();
				logger.severe("Can't draw surface map: " + e);
			}
		}
	}

```

**Comment by mokun** [comment]
> I think I can make it work. Key is to add a isReady method to the MapData interface.

You can say calling drawMap can be put on hold easily by just not executing the thread.

Also, there is a boolean called recreateMap plus other criteria in MapPanel itself for delaying the map update. 

So now it's a matter of integrating the old mechanism with the new isReady method to await for a new map to be downloaded so as to reconstruct a new pixel array in `IntegerMapData` for a new rendering of a projection of Mars map.

```java
	/**
	 * Displays map at given center, regenerating if necessary.
	 *
	 * @param newCenter the center location for the globe
	 * @param rho
	 */
	public void showMap(Coordinates newCenter, double rho) {
		if (newCenter == null) 
			return;
		
		if (centerCoords == null
			|| rho != getRho()
			|| !centerCoords.equals(newCenter)
			) {
				recreateMap = true;
				centerCoords = newCenter;
		}
		else {
			recreateMap = false;
		}
			
		if (recreateMap) {
			wait = true;
			updateDisplay(rho);
			recreateMap = false;
		}
	}

```
@muybien3433, just FYI.

**Comment by bevans2000** [comment]
Yes it would work but it means the map display would not update when the download is running in the background.
But I think the way to do this is to keep a hold of the MapData loading on the background and check it for isReady when drawing. So it automatically flips to the new map layer once it is ready.
Hence the sync callback is only needed between the file loading and the MapData caller. I've nearly got that working

**Comment by bevans2000** [comment]
I think I've got this working now. Just wanted to test it and clean it up. Very busy at work so not sure how much time I'll have before the weekend.

**Comment by bevans2000** [comment]
@mokun This map logic is a bit of a mess. There are static attributes in the IntegerMapData class that are used in the MapPanel that get updated everytime a map loads. These seem to be field attributes by stealth; really bad design.
I've got to change this because this breaks down when a 2nd map is loaded in the background.

**Comment by mokun** [comment]
> what does the value 'rho' represent in the map logic? 

*rho* is simply the radius of Mars in terms of # of pixels. 

Since the width of the image is also 2 x pi x *rho* (the circumference of Mars), *rho* = # of pixels of image width / 2 / pi.

> Is it something about the scale that is being applied to the display? 

scale = rho / rho_default

In MapPanel, we change the magnification (or the scale) of the map to be displayed as the player rolls the mouse wheel.

**Comment by mokun** [comment]
The initial rho (rho_default) was determined by the map image width.

But since the zoom slider is in MapPanel, when the mouse wheel rolls, it changes the value of rho.

So rho can be changed by player or by the map being selected.

> The scale should be pass in as an argument when doing the rendering?

It seems easier to track rho instead of scale. I've tried.

The thing is that the convertRectIntToSpherical() method in IntegerMapData or its equivalent formula in OpenCL uses rho as a param, not scale.

**Comment by bevans2000** [comment]
Ok, so rho is really used to scale the image. The default_rho is driven by the image selected, I get it now.
So the max, min and default rho should be properties of the MapData instead of static. If I do this I think it will fix the scale problem when loading maps asynchronously. Everything should be driven by the map data on display

**Comment by bevans2000** [comment]
They way I will redo this is to keep the current rho value in the Map display and not the Map data. The latter should be ready only once loaded and provide a stable set of images according to what the display wants.

**Comment by mokun** [comment]
The interesting thing is that BOTH the zoom slider and the loading of a new map image would drive the value of rho. 

It's really great that you've found a way to simplify tracking rho

**Comment by bevans2000** [comment]
> It's really great that you've found a way to simplify tracking rho

Hang on I'm not there yet 😀

**Comment by bevans2000** [comment]
So it's working but I have slightly had to change how the zoom is calculated. It is now totally driven by the height of the underlying image which makes sense as that is the width for the Mars globe. This would then be the minimum RHO.
But for the higher resolution maps, I cannot zoom out to show the full globe which is a puzzle. 
I've got to add some decoration so its obvious there is a background load happening.

**Comment by mokun** [comment]
> But for the higher resolution maps, I cannot zoom out to show the full globe which is a puzzle.

That depends on how you define the range of the "scale"  of the zoom slider.

Right now, I define the scale each time a new map image is loaded. The scale is always from something like 1/5 to 5 (could be 1/6 to 6), no matter what rho is.

I did it this way in order to preserve the value of rho seamlessly

The value of the scale has to change each time in order to zoom in and out for the new rho to match the previous rho.

This way, the map won't abruptly "jump".

**Comment by bevans2000** [comment]
I have gone back to look how it works on 3.8 and it seems to behave in the same way as my new version. The minimum rho has to derived from the height of the image loaded, e.g at the minimum zoom scaler the whole globe should be visible. 
I might clean it up and commit it to the branch and you can have a go with it.

**Comment by mokun** [comment]
The old way won't give you seamless transition from one map to the next though. It always abruptly jump since the value of rho did not get preserved.

I don't remember if 3.8.0 has seamless transition or after that.

**Comment by bevans2000** [comment]
Right, no 3.8 is not seamless. So you want rho to be preserved when jumping between maps. So the value of the zoom slider may change when the map changes? Each map may have a different min and max rho.
At the moment, I've got it the same as 3.8 where the slider stays the same but the rho changes when map is flipped.
So the zoom slider value represents the relative position of the current rho between the min and max for that map. I think I understand how you want it to work. Should be possible to do that with this cleaner new approach.
If I can find time I will commit a working version to the branch tomorrow.

**Comment by mokun** [comment]
> It is now totally driven by the height of the underlying image which makes sense as that is the width for the Mars globe.

For a global map, the width of the image the player wants to project has to be half its height.

> w = 2 * h

> 2 * rho * PI = w

> rho = w / 2 / PI
 
> rho = h / PI

But in future, we will have a sectional map or a tile map, the w and h will NOT have the same ratio, you know.

**Comment by mokun** [comment]
> So you want rho to be preserved when jumping between maps. So the value of the zoom slider may change when the map changes? Each map may have a different min and max rho.

Yeah currently, rho would vary and is NOT fixed. So the max rho (the upper limit) and min rho (the lower limit) on each map would have to be different.

rho is the radius of Mars is just the # of pixels of the Masr Globe that the player wants to display as he zooms in and zooms out. 

So rho varies with the mouse wheel.

That's why rho could vary from whatever it's lowest value is, say 100 pixels (at lowest resolution), to, say 10,000 pixels (at the highest resolution).

**Comment by bevans2000** [comment]
So I think what you describe is how it is working. I will check later

**Comment by mokun** [comment]
> will have a sectional map or a tile map, the w and h will NOT have the same ratio, you know.

I see what's next is to allow loading of a regional map. Say it's a very small region in Valles Marineris. A 1200 x 1200 pixels  showcases the terrain with incredible detail. 

The meta data that comes with this regional map is important. 

Apparently, it will need to define at least (1) the center coordinate of this map; and (2) the resolution/scale of this map.

Based on these two pieces of info, mars-sim will select an appropriate surface map of a certain resolution level as its background global map upon which this regional map will superimpose.

There could be hundreds of these regional maps that we pre-select and made available for download in Github.

**Comment by bevans2000** [comment]
Yes I agree. I think the changes I have done abstract more of the underlying map data from the UI. 
I'm still thinking about whether rho is the best way to handle scaling but not going to change it yet. I wonder if a simple meters/pixel ratio might be easier to understand

**Comment by mokun** [comment]
> the changes I have done abstract more of the underlying map data from the UI.

So can you move IntegerMapData class to the ui package ?

The console version of mars-sim has no need of reading a map image and reassemble it to create a map projection.

**Comment by mokun** [comment]
@bevans2000 @Urwumpe 

I just watched [this](https://www.youtube.com/watch?v=xC8qfXxAhAw) and it dawns on me that we are supposed to ensure it will connect to GitHub raw site only when downloading designated maps. 

Or else, it can become a vulnerability down the road.

The question is, what measures would you put in place ?

**Comment by bevans2000** [comment]
Checksum is the right way to go. But the range of map layers is configuration file driven, so baking the checksums in the fixed code wouldn't work.
The only way would be to have the map ranges in the code as well. Very possible.
However, the files downloaded are pushed to disk first and then consumed as images by the simulation. 
Local machine protection should detect a bad file on the local drive before the simulation opens it or it would prevent the Java executable downloading files.
The other alternative is we have the extra maps as a pack or add-on that is downloaded once. This would be big but as it's a single file we could have the checksum for that hard coded.

**Comment by bevans2000** [comment]
In addition, its not perfect then user downloads this from GitHub as it's a set of big files in the main repo. Ideally we should move them to a separate repo and then host on the website [marssim](https://marssim.space) .. although that is still GitHub pages at the moment as well
